### PR TITLE
Support Spark-independent ways to upload data to TD under a unified interface

### PR DESCRIPTION
**Pull request by takuti**
As of v0.3.0, pytd deterministically uses Plazma API (via td-spark) to upload data frame to TD. However, we do have many different ways to upload client-side data to TD, and we can alternatively use local Embulk instance or INSERT INTO query statement. Note that optimal choice may vary depending on data size, network limitation, etc. Plus, Plazma API is still experimental, and it introduces extra dependency to Java runtime.

Therefore, pytd should support the other ways to upload data in addition to the current Plazma API’s one, so that users can flexibly (or, pytd itself can dynamically) choose a reasonable option. Regardless of the choice of upload method, API should be unique for usability.

### CODE REVIEW

**Comment by takuti** [comment]
@chezou Made some decisions to change the interfaces. Could you take a look at the modifications to see if it looks reasonable? 

Here is sample code you might want to try:

```py
import pytd
from td_pyspark import TDSparkContextBuilder

import pandas as pd
import numpy as np
import datetime

database, table = 'takuti', 'foo'

client = pytd.Client(database=database, default_engine='hive')

df = pd.DataFrame({'A' : ['spam', 'eggs', 'spam', 'eggs'] * 6,
                   'B' : [None] + [['alpha', 'beta', 'gamma']] * 23,
                   'C' : [np.random.choice(pd.date_range(datetime.datetime(2013,1,1),datetime.datetime(2013,1,3))) for i in range(24)],
                   'D' : np.random.randn(24),
                   'E' : np.random.random_integers(0,4,24)})

# Option 1: INSERT INTO query
client.load_table_from_dataframe(df, '{}.{}'.format(database, table), writer='insert_into', if_exists='overwrite')

# Option 2: Bulk import API
client.load_table_from_dataframe(df, '{}.{}'.format(database, table), if_exists='overwrite')

# Option 3a: Spark writer
client.load_table_from_dataframe(df, '{}.{}'.format(database, table), writer='spark', if_exists='append')

# Option 3b: Spark writer with custom td-spark
jar_path = TDSparkContextBuilder.default_jar_path()
client.load_table_from_dataframe(df, '{}.{}'.format(database, table), writer='spark', if_exists='append', td_spark_path=jar_path)

# Table#import_dataframe; this should fail
client.get_table(database, table).import_dataframe(df, writer='insert_into', if_exists='error')

# Deleting table
tbl = client.get_table(database, table)
tbl.delete()

# Table#import_dataframe with pre-defined, reusable Writer instance
from pytd.writer import SparkWriter
jar_path = TDSparkContextBuilder.default_jar_path()
writer = SparkWriter(td_spark_path=jar_path)
tbl.import_dataframe(df, writer=writer, if_exists='error')

# Writer#write_dataframe; this should fail
writer.write_dataframe(df, tbl, if_exists='append')
```

**Comment by chezou** [comment]
@takuti It looks reasonable for me to close the writer when the `writer` option is a string. And understood that `load_table_from_dataframe` function is a shortcut of `tbl.import_dataframe`.

Will look into more in detail.

File: pytd/tests/test_writer.py [hunk 1]
```diff
@@ -3,25 +3,124 @@
...
+        # 3) INSERT INTO
+        self.assertEqual(self.table.client.query.call_count, 1)
+
+    def test_write_dataframe_invalid_if_exists(self):
+        with self.assertRaises(ValueError):
+            self.writer.write_dataframe(
+                pd.DataFrame([[1, 2], [3, 4]]), self.table, if_exists="bar"
+            )
+
+    def test_close(self):
```
[the remark this answers is not in the document]
Comment (takuti) [reply pytd/tests/test_writer.py]: Just for the runnability. But changed to make it more meaningful: c1d3023e4243139b18bacea21b781798e800cd69

**Comment by takuti** [comment]
Thank you for the review. Let me spend some time to think about the type issue you pointed out.

**Comment by chezou** [comment]
As you may know, here is the list of dtypes of pandas DataFrame which may be used as input https://pandas.pydata.org/pandas-docs/stable/getting_started/basics.html#dtypes

**Comment by takuti** [comment]
@chezou PTAL

Added dtype normalization logic to overcome the type difference issue: 65cd7cf51e9e57fe7f76d966a9363e996b86f79b

For array type support, since it introduces extra complexity e.g., escaping quotes, normalizing type of array elements, I just created a new issue #28 for now. Let's have further consideration later.

File: pytd/writer.py [hunk 2]
```diff
@@ -2,36 +2,309 @@
...
+    f  floating-point
+    c  complex floating-point
+    m  timedelta
+    M  datetime
+    O  object
+    S  (byte-)string
+    U  Unicode
+    V  void
+    """
+    df = dataframe.copy()
```
Comment (chezou) [inline pytd/writer.py]: Just confirming. Does this line mean copy for every conversion? Don't we need some option like `inplace?`

Comment (takuti) [reply pytd/writer.py]: Agreed. eb966de58698d1d7cbcadb7aec7cfd99d586116e

**Pull request merged**